File: android/shared/src/commonMain/kotlin/com/sitepinapp/services/ProjectSharingService.kt
package com.sitepinapp.services

import com.sitepinapp.data.model.*
import com.sitepinapp.data.repository.SitePinRepository
import com.sitepinapp.platform.ImportLimits
import kotlinx.datetime.Clock
import kotlinx.serialization.json.Json
import kotlin.math.abs
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid

@OptIn(ExperimentalUuidApi::class)
object ProjectSharingService {

    private val json = Json { prettyPrint = true; ignoreUnknownKeys = true }

    private const val MAX_STRING_LENGTH = 1000
    private const val MAX_DESCRIPTION_LENGTH = 5000
    private const val MAX_FORMAT_VERSION = 2
    private val validStatuses = setOf("open", "in_progress", "resolved")
    private val validCategories = PinCategory.entries.map { it.name.lowercase() }.toSet()

    private fun sanitize(value: String, maxLength: Int = MAX_STRING_LENGTH): String = value.take(maxLength)
    private fun sanitizeStatus(value: String): String = if (value in validStatuses) value else "open"
    private fun sanitizeCategory(value: String): String = if (value in validCategories) value else "general"
    private fun sanitizeSyncID(value: String): String = value.trim().ifEmpty { Uuid.random().toString() }.take(100)

    // ── EXPORT ──

    suspend fun exportProjectToJson(repository: SitePinRepository, projectId: Long): String {
        val fullData = repository.getFullProjectData(projectId)
            ?: throw IllegalStateException("Project not found")

        val export = ProjectExport(
            name = fullData.project.name,
            syncID = fullData.project.syncID,
            createdAt = DateUtils.toISO8601(fullData.project.createdAt),
            exportedBy = UserProfileManager.getDisplayName(),
            exportedAt = DateUtils.toISO8601(Clock.System.now().toEpochMilliseconds()),
            documents = fullData.documents.map { docData ->
                DocumentExport(
                    name = docData.document.name,
                    syncID = docData.document.syncID,
                    fileType = docData.document.fileType,
                    pageCount = docData.document.pageCount,
                    createdAt = DateUtils.toISO8601(docData.document.createdAt),
                    fileData = DateUtils.encodeBase64(docData.document.fileData),
                    pins = docData.pins.map { pinData ->
                        PinExport(
                            syncID = pinData.pin.syncID,
                            relativeX = pinData.pin.relativeX,
                            relativeY = pinData.pin.relativeY,
                            pageIndex = pinData.pin.pageIndex,
                            title = pinData.pin.title,
                            pinDescription = pinData.pin.pinDescription,
                            location = pinData.pin.location,
                            height = pinData.pin.height,
                            width = pinData.pin.width,
                            status = pinData.pin.status,
                            category = pinData.pin.category,
                            author = pinData.pin.author,
                            createdAt = DateUtils.toISO8601(pinData.pin.createdAt),
                            modifiedAt = DateUtils.toISO8601(pinData.pin.modifiedAt),
                            photos = pinData.photos.map { PhotoExport(DateUtils.encodeBase64(it.imageData), it.caption, DateUtils.toISO8601(it.createdAt), it.syncID) },
                            comments = pinData.comments.map { CommentExport(it.text, it.author, DateUtils.toISO8601(it.createdAt), it.syncID) }
                        )
                    }
                )
            }
        )
        return json.encodeToString(ProjectExport.serializer(), export)
    }

    // ── IMPORT ──

    suspend fun importProjectFromJson(repository: SitePinRepository, jsonString: String): Long {
        // A4: Validate import size
        require(jsonString.length <= ImportLimits.MAX_IMPORT_SIZE_BYTES) {
            "Import file is too large (${jsonString.length / (1024 * 1024)} MB). Maximum is ${ImportLimits.MAX_IMPORT_SIZE_BYTES / (1024 * 1024)} MB."
        }

        val export = json.decodeFromString(ProjectExport.serializer(), jsonString)

        // Validate format version
        require(export.formatVersion <= MAX_FORMAT_VERSION) {
            "This file was created with a newer version of SitePin (format v${export.formatVersion}). Please update the app."
        }

        // A3: Validate photo counts per pin
        for (docExport in export.documents) {
            for (pinExport in docExport.pins) {
                require(pinExport.photos.size <= ImportLimits.MAX_PHOTOS_PER_PIN) {
                    "Pin \"${pinExport.title.ifEmpty { "Untitled" }}\" has ${pinExport.photos.size} photos, exceeding the maximum of ${ImportLimits.MAX_PHOTOS_PER_PIN}."
                }
            }
        }

        val project = Project(name = sanitize(export.name), syncID = sanitizeSyncID(export.syncID), createdAt = DateUtils.fromISO8601(export.createdAt))
        val projectId = repository.insertProject(project)

        for (docExport in export.documents) {
            val document = PlanDocument(
                projectId = projectId, name = sanitize(docExport.name), syncID = sanitizeSyncID(docExport.syncID),
                fileType = docExport.fileType, pageCount = docExport.pageCount,
                createdAt = DateUtils.fromISO8601(docExport.createdAt),
                fileData = DateUtils.decodeBase64(docExport.fileData)
            )
            val documentId = repository.insertDocument(document)

            for (pinExport in docExport.pins) {
                val pin = Pin(
                    documentId = documentId, syncID = sanitizeSyncID(pinExport.syncID),
                    relativeX = pinExport.relativeX.coerceIn(0.0, 1.0),
                    relativeY = pinExport.relativeY.coerceIn(0.0, 1.0),
                    pageIndex = pinExport.pageIndex, title = sanitize(pinExport.title),
                    pinDescription = sanitize(pinExport.pinDescription, MAX_DESCRIPTION_LENGTH),
                    location = sanitize(pinExport.location),
                    height = sanitize(pinExport.height), width = sanitize(pinExport.width),
                    status = sanitizeStatus(pinExport.status), category = sanitizeCategory(pinExport.category),
                    author = sanitize(pinExport.author),
                    createdAt = DateUtils.fromISO8601(pinExport.createdAt),
                    modifiedAt = DateUtils.fromISO8601(pinExport.modifiedAt)
                )
                val pinId = repository.insertPin(pin)

                for (photoExport in pinExport.photos) {
                    val photoSyncID = sanitizeSyncID(photoExport.syncID)
                    repository.insertPhoto(PinPhoto(pinId = pinId, imageData = DateUtils.decodeBase64(photoExport.imageData), caption = sanitize(photoExport.caption), createdAt = DateUtils.fromISO8601(photoExport.createdAt), syncID = photoSyncID))
                }
                for (commentExport in pinExport.comments) {
                    val commentSyncID = sanitizeSyncID(commentExport.syncID)
                    repository.insertComment(PinComment(pinId = pinId, text = sanitize(commentExport.text, MAX_DESCRIPTION_LENGTH), author = sanitize(commentExport.author), createdAt = DateUtils.fromISO8601(commentExport.createdAt), syncID = commentSyncID))
                }
            }
        }
        return projectId
    }

    // ── MERGE ──

    suspend fun mergeProjectFromJson(repository: SitePinRepository, jsonString: String, existingProjectId: Long) {
        // Validate import size and format
        require(jsonString.length <= ImportLimits.MAX_IMPORT_SIZE_BYTES) {
            "Import file is too large (${jsonString.length / (1024 * 1024)} MB). Maximum is ${ImportLimits.MAX_IMPORT_SIZE_BYTES / (1024 * 1024)} MB."
        }
        val export = json.decodeFromString(ProjectExport.serializer(), jsonString)
        require(export.formatVersion <= MAX_FORMAT_VERSION) {
            "This file was created with a newer version of SitePin (format v${export.formatVersion}). Please update the app."
        }
        // Validate photo counts per pin
        for (docExport in export.documents) {
            for (pinExport in docExport.pins) {
                require(pinExport.photos.size <= ImportLimits.MAX_PHOTOS_PER_PIN) {
                    "Pin \"${pinExport.title.ifEmpty { "Untitled" }}\" has ${pinExport.photos.size} photos, exceeding the maximum of ${ImportLimits.MAX_PHOTOS_PER_PIN}."
                }
            }
        }

        for (docExport in export.documents) {
            val existingDoc = repository.getDocumentBySyncID(existingProjectId, docExport.syncID)
            val documentId: Long = existingDoc?.id ?: repository.insertDocument(
                PlanDocument(
                    projectId = existingProjectId, name = sanitize(docExport.name), syncID = sanitizeSyncID(docExport.syncID),
                    fileType = docExport.fileType, pageCount = docExport.pageCount,
                    createdAt = DateUtils.fromISO8601(docExport.createdAt),
                    fileData = DateUtils.decodeBase64(docExport.fileData)
                )
            )

            for (pinExport in docExport.pins) {
                val existingPin = repository.getPinBySyncID(documentId, pinExport.syncID)
                val pinId: Long

                if (existingPin != null) {
                    pinId = existingPin.id
                    val incomingModified = DateUtils.fromISO8601(pinExport.modifiedAt)
                    if (incomingModified > existingPin.modifiedAt) {
                        repository.updatePin(existingPin.copy(
                            title = sanitize(pinExport.title),
                            pinDescription = sanitize(pinExport.pinDescription, MAX_DESCRIPTION_LENGTH),
                            location = sanitize(pinExport.location),
                            height = sanitize(pinExport.height), width = sanitize(pinExport.width),
                            status = sanitizeStatus(pinExport.status), category = sanitizeCategory(pinExport.category),
                            modifiedAt = incomingModified
                        ))
                    }
                } else {
                    pinId = repository.insertPin(Pin(
                        documentId = documentId, syncID = sanitizeSyncID(pinExport.syncID),
                        relativeX = pinExport.relativeX.coerceIn(0.0, 1.0),
                        relativeY = pinExport.relativeY.coerceIn(0.0, 1.0),
                        pageIndex = pinExport.pageIndex, title = sanitize(pinExport.title),
                        pinDescription = sanitize(pinExport.pinDescription, MAX_DESCRIPTION_LENGTH),
                        location = sanitize(pinExport.location),
                        height = sanitize(pinExport.height), width = sanitize(pinExport.width),
                        status = sanitizeStatus(pinExport.status), category = sanitizeCategory(pinExport.category),
                        author = sanitize(pinExport.author),
                        createdAt = DateUtils.fromISO8601(pinExport.createdAt),
                        modifiedAt = DateUtils.fromISO8601(pinExport.modifiedAt)
                    ))
                }

                // Merge comments (use syncID if available, fall back to text+author+timestamp)
                val existingComments = repository.getCommentsByPinIdSuspend(pinId)
                val commentList: List<CommentExport> = pinExport.comments.toList()
                for (c: CommentExport in commentList) {
                    val cTime = DateUtils.fromISO8601(c.createdAt)
                    val isDup = if (c.syncID.isNotBlank()) {
                        existingComments.any { it.syncID == c.syncID }
                    } else {
                        existingComments.any { it.text == c.text && it.author == c.author && abs(it.createdAt - cTime) < 1000 }
                    }
                    if (!isDup) {
                        val syncID = sanitizeSyncID(c.syncID)
                        repository.insertComment(PinComment(pinId = pinId, text = sanitize(c.text, MAX_DESCRIPTION_LENGTH), author = sanitize(c.author), createdAt = cTime, syncID = syncID))
                    }
                }

                // Merge photos (use syncID if available, fall back to caption+timestamp)
                val existingPhotos = repository.getPhotosByPinIdSuspend(pinId)
                val photoList: List<PhotoExport> = pinExport.photos.toList()
                for (p: PhotoExport in photoList) {
                    val pTime = DateUtils.fromISO8601(p.createdAt)
                    val isDup = if (p.syncID.isNotBlank()) {
                        existingPhotos.any { it.syncID == p.syncID }
                    } else {
                        existingPhotos.any { it.caption == p.caption && abs(it.createdAt - pTime) < 1000 }
                    }
                    if (!isDup) {
                        val syncID = sanitizeSyncID(p.syncID)
                        repository.insertPhoto(PinPhoto(pinId = pinId, imageData = DateUtils.decodeBase64(p.imageData), caption = sanitize(p.caption), createdAt = pTime, syncID = syncID))
                    }
                }
            }
        }
    }
}

File: ios/SitePinApp/Services/ProjectSharingService.swift
import Foundation
import SwiftData

struct ProjectSharingService {

    // MARK: - Export

    static func exportProject(_ project: Project) throws -> Data {
        let export = ProjectExport(
            name: project.name,
            syncID: project.syncID,
            exportedBy: UserProfileManager.shared.displayName,
            exportedAt: Date(),
            documents: project.documents.sorted(by: { $0.createdAt < $1.createdAt }).map { doc in
                DocumentExport(
                    name: doc.name,
                    syncID: doc.syncID,
                    fileData: doc.fileData,
                    fileType: doc.fileType,
                    pageCount: doc.pageCount,
                    createdAt: doc.createdAt,
                    pins: doc.pins.sorted(by: { $0.createdAt < $1.createdAt }).map { pin in
                        PinExport(
                            syncID: pin.syncID,
                            relativeX: pin.relativeX,
                            relativeY: pin.relativeY,
                            title: pin.title,
                            pinDescription: pin.pinDescription,
                            location: pin.location,
                            height: pin.height,
                            width: pin.width,
                            status: pin.status,
                            category: pin.category,
                            author: pin.author,
                            pageIndex: pin.pageIndex,
                            createdAt: pin.createdAt,
                            modifiedAt: pin.modifiedAt,
                            photos: pin.photos.map { photo in
                                PhotoExport(imageData: photo.imageData, caption: photo.caption, createdAt: photo.createdAt, syncID: photo.syncID)
                            },
                            comments: pin.comments.sorted(by: { $0.createdAt < $1.createdAt }).map { comment in
                                CommentExport(text: comment.text, author: comment.author, createdAt: comment.createdAt, syncID: comment.syncID)
                            }
                        )
                    }
                )
            }
        )

        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        return try encoder.encode(export)
    }

    static func exportToFile(_ project: Project) throws -> URL {
        let data = try exportProject(project)
        let safeName = sanitizeFileName(project.name)
        let fileName = "\(safeName).sitepin"
        let url = FileManager.default.temporaryDirectory.appendingPathComponent(fileName)
        try data.write(to: url)
        return url
    }

    // MARK: - Import

    enum ImportError: LocalizedError {
        case securityScopedAccessDenied
        case fileTooLarge(sizeMB: Int, maxMB: Int)
        case photoLimitExceeded(pinTitle: String, count: Int, max: Int)

        var errorDescription: String? {
            switch self {
            case .securityScopedAccessDenied:
                return "Could not access the selected file. Please try again."
            case .fileTooLarge(let sizeMB, let maxMB):
                return "File is too large (\(sizeMB) MB). Maximum allowed is \(maxMB) MB."
            case .photoLimitExceeded(let pinTitle, let count, let max):
                return "Pin \"\(pinTitle)\" has \(count) photos, exceeding the maximum of \(max)."
            }
        }
    }

    static func importProject(from url: URL, into context: ModelContext) throws -> Project {
        guard url.startAccessingSecurityScopedResource() else {
            throw ImportError.securityScopedAccessDenied
        }
        defer { url.stopAccessingSecurityScopedResource() }

        let data = try Data(contentsOf: url)
        return try importProject(from: data, into: context)
    }

    static func importProject(from data: Data, into context: ModelContext) throws -> Project {
        // A3/A4: Validate file size before parsing
        if data.count > SitePinConstants.maxImportSize {
            throw ImportError.fileTooLarge(sizeMB: data.count / (1024 * 1024), maxMB: SitePinConstants.maxImportSize / (1024 * 1024))
        }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        let export = try decoder.decode(ProjectExport.self, from: data)

        // A3: Validate photo counts per pin before importing
        for doc in export.documents {
            for pin in doc.pins where pin.photos.count > SitePinConstants.maxPhotosPerPin {
                throw ImportError.photoLimitExceeded(
                    pinTitle: pin.title.isEmpty ? "Untitled" : pin.title,
                    count: pin.photos.count,
                    max: SitePinConstants.maxPhotosPerPin
                )
            }
        }

        let project = Project(name: sanitizeString(export.name))
        project.syncID = sanitizeSyncID(export.syncID)
        context.insert(project)

        for docExport in export.documents {
            let doc = PlanDocument(name: sanitizeString(docExport.name), fileData: docExport.fileData,
                                   fileType: docExport.fileType, pageCount: docExport.pageCount)
            doc.syncID = sanitizeSyncID(docExport.syncID)
            doc.createdAt = docExport.createdAt
            doc.project = project
            project.documents.append(doc)

            for pinExport in docExport.pins {
                let pin = Pin(relativeX: pinExport.relativeX, relativeY: pinExport.relativeY,
                              pageIndex: pinExport.pageIndex)
                pin.syncID = sanitizeSyncID(pinExport.syncID)
                applyPinExportToPin(pinExport, pin: pin)
                pin.createdAt = pinExport.createdAt
                pin.modifiedAt = pinExport.modifiedAt
                pin.document = doc
                doc.pins.append(pin)

                for photoExport in pinExport.photos {
                    let photo = PinPhoto(imageData: photoExport.imageData, caption: sanitizeString(photoExport.caption))
                    photo.createdAt = photoExport.createdAt
                    if let sid = photoExport.syncID, !sid.isEmpty { photo.syncID = sanitizeSyncID(sid) }
                    photo.pin = pin
                    pin.photos.append(photo)
                }

                for commentExport in pinExport.comments {
                    let comment = PinComment(text: sanitizeString(commentExport.text, maxLength: maxDescriptionLength),
                                             author: sanitizeString(commentExport.author), preserveEmptyAuthor: true)
                    comment.createdAt = commentExport.createdAt
                    if let sid = commentExport.syncID, !sid.isEmpty { comment.syncID = sanitizeSyncID(sid) }
                    comment.pin = pin
                    pin.comments.append(comment)
                }
            }
        }

        return project
    }

    // MARK: - Sync / Merge

    static func mergeProject(from url: URL, into existingProject: Project, context: ModelContext) throws {
        guard url.startAccessingSecurityScopedResource() else {
            throw ImportError.securityScopedAccessDenied
        }
        defer { url.stopAccessingSecurityScopedResource() }

        let data = try Data(contentsOf: url)

        // A4: Validate file size
        if data.count > SitePinConstants.maxImportSize {
            throw ImportError.fileTooLarge(sizeMB: data.count / (1024 * 1024), maxMB: SitePinConstants.maxImportSize / (1024 * 1024))
        }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        let export = try decoder.decode(ProjectExport.self, from: data)

        for docExport in export.documents {
            if let existingDoc = existingProject.documents.first(where: { $0.syncID == docExport.syncID }) {
                for pinExport in docExport.pins {
                    if let existingPin = existingDoc.pins.first(where: { $0.syncID == pinExport.syncID }) {
                        let incomingModified = pinExport.modifiedAt
                        if incomingModified > existingPin.modifiedAt {
                            applyPinExportToPin(pinExport, pin: existingPin)
                            existingPin.modifiedAt = incomingModified
                        }

                        // Merge new comments (use syncID if available, fall back to text+author+timestamp)
                        for commentExport in pinExport.comments {
                            let alreadyExists: Bool
                            if let sid = commentExport.syncID, !sid.isEmpty {
                                alreadyExists = existingPin.comments.contains(where: { $0.syncID == sid })
                            } else {
                                alreadyExists = existingPin.comments.contains(where: {
                                    $0.text == commentExport.text && $0.author == commentExport.author &&
                                    abs($0.createdAt.timeIntervalSince(commentExport.createdAt)) < 1
                                })
                            }
                            if !alreadyExists {
                                let comment = PinComment(text: sanitizeString(commentExport.text, maxLength: maxDescriptionLength),
                                                         author: sanitizeString(commentExport.author), preserveEmptyAuthor: true)
                                comment.createdAt = commentExport.createdAt
                                if let sid = commentExport.syncID, !sid.isEmpty { comment.syncID = sanitizeSyncID(sid) }
                                comment.pin = existingPin
                                existingPin.comments.append(comment)
                            }
                        }

                        // Merge new photos (use syncID if available, fall back to caption+timestamp)
                        for photoExport in pinExport.photos {
                            let alreadyExists: Bool
                            if let sid = photoExport.syncID, !sid.isEmpty {
                                alreadyExists = existingPin.photos.contains(where: { $0.syncID == sid })
                            } else {
                                alreadyExists = existingPin.photos.contains(where: {
                                    $0.caption == photoExport.caption &&
                                    abs($0.createdAt.timeIntervalSince(photoExport.createdAt)) < 1
                                })
                            }
                            if !alreadyExists {
                                let photo = PinPhoto(imageData: photoExport.imageData, caption: sanitizeString(photoExport.caption))
                                photo.createdAt = photoExport.createdAt
                                if let sid = photoExport.syncID, !sid.isEmpty { photo.syncID = sanitizeSyncID(sid) }
                                photo.pin = existingPin
                                existingPin.photos.append(photo)
                            }
                        }
                    } else {
                        // New pin
                        addImportedPin(pinExport, to: existingDoc)
                    }
                }
            } else {
                // New document
                let doc = PlanDocument(name: sanitizeString(docExport.name), fileData: docExport.fileData,
                                       fileType: docExport.fileType, pageCount: docExport.pageCount)
                doc.syncID = sanitizeSyncID(docExport.syncID)
                doc.createdAt = docExport.createdAt
                doc.project = existingProject
                existingProject.documents.append(doc)

                for pinExport in docExport.pins {
                    addImportedPin(pinExport, to: doc)
                }
            }
        }

    }

    // MARK: - Input Sanitization

    private static let maxStringLength = 1000
    private static let maxDescriptionLength = 5000
    private static let validStatuses = Set(PinStatus.allCases.map(\.rawValue))
    private static let validCategories = Set(PinCategory.allCases.map(\.rawValue))

    private static func sanitizeString(_ value: String, maxLength: Int = maxStringLength) -> String {
        String(value.prefix(maxLength))
    }

    private static func sanitizeStatus(_ value: String) -> String {
        validStatuses.contains(value) ? value : PinStatus.open.rawValue
    }

    private static func sanitizeCategory(_ value: String) -> String {
        validCategories.contains(value) ? value : PinCategory.general.rawValue
    }

    private static func sanitizeSyncID(_ value: String) -> String {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? UUID().uuidString : String(trimmed.prefix(100))
    }

    private static func applyPinExportToPin(_ pinExport: PinExport, pin: Pin) {
        pin.title = sanitizeString(pinExport.title)
        pin.pinDescription = sanitizeString(pinExport.pinDescription, maxLength: maxDescriptionLength)
        pin.location = sanitizeString(pinExport.location)
        pin.height = sanitizeString(pinExport.height)
        pin.width = sanitizeString(pinExport.width)
        pin.status = sanitizeStatus(pinExport.status)
        pin.category = sanitizeCategory(pinExport.category)
        pin.author = sanitizeString(pinExport.author)
    }

    // MARK: - Helpers

    private static func addImportedPin(_ pinExport: PinExport, to doc: PlanDocument) {
        let pin = Pin(relativeX: pinExport.relativeX, relativeY: pinExport.relativeY,
                      pageIndex: pinExport.pageIndex)
        pin.syncID = sanitizeSyncID(pinExport.syncID)
        applyPinExportToPin(pinExport, pin: pin)
        pin.createdAt = pinExport.createdAt
        pin.modifiedAt = pinExport.modifiedAt
        pin.document = doc
        doc.pins.append(pin)

        for photoExport in pinExport.photos {
            let photo = PinPhoto(imageData: photoExport.imageData, caption: sanitizeString(photoExport.caption))
            photo.createdAt = photoExport.createdAt
            if let sid = photoExport.syncID, !sid.isEmpty { photo.syncID = sanitizeSyncID(sid) }
            photo.pin = pin
            pin.photos.append(photo)
        }
        for commentExport in pinExport.comments {
            let comment = PinComment(text: sanitizeString(commentExport.text, maxLength: maxDescriptionLength),
                                     author: sanitizeString(commentExport.author), preserveEmptyAuthor: true)
            comment.createdAt = commentExport.createdAt
            if let sid = commentExport.syncID, !sid.isEmpty { comment.syncID = sanitizeSyncID(sid) }
            comment.pin = pin
            pin.comments.append(comment)
        }
    }

    private static func sanitizeFileName(_ name: String) -> String {
        let sanitized = name.replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "\\", with: "_")
            .replacingOccurrences(of: ":", with: "_")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        return sanitized.isEmpty ? "Untitled" : String(sanitized.prefix(100))
    }
}
